### ISSUE TITLE
Magento not actioning all url parameters on Product Page if a swatch/color option  is included

### ISSUE THREAD

**Issue by UKRogue** [magento/magento2#19724]
### Preconditions (*)
Magento 2.2.6 or Magento 2.2.3
PHP 7
Porto or Magento Blank theme

### Steps to reproduce (*)
1. Create 2 product attributes 1 for a swatch color, the other a height attribute, add a minimum of 2 options for each. ie Color Red and Blue, Height 2mm and 4mm.
2. Create 4 simple products with differant images and prices
a) product 1 with options Red, 2mm price: £2
b) product 2 with options Red, 4mm price: £3
c) product 3 with options Blue, 2mm price: £4
d) product 4 with options Blue, 4mm price: £5

3. Create a configurable with color and height options, add the 4 simple products to the configurable.
4.  For the sake of argument lets say that the created options have id's of Color:1 and Height: 2 and each option has a reference 1 and 2,
5. Use the configurable url with parameters to  select any of the simple products ie https://domainname.co.uk#1=2&2=2 (Change parameters to suit your options)

### Expected result (*)
1. When the product page loads, Magento should identify the url parameters and both options should be selected, the image and price should change to reflect the option(s) in the url parameters.

![image](https://user-images.githubusercontent.com/22025506/49872449-fe0d0c80-fe10-11e8-9bf1-91f8461b3894.png)

### Actual result (*)

1. If a swatch is included in the option and another option exists (In this case packaging) Only the swatch colour is selected. The second option is ignored. Image is always loaded correctly.
2. If there is no swatch as an option, all parameters are actioned correctly and the price and image changed to match
![image](https://user-images.githubusercontent.com/22025506/49872464-07967480-fe11-11e8-8d6b-9d92cd167964.png)

This causes me a problem as any landing page with a swatch that has other options is reported incorrectly to google shoppping from my merchant feed which generates the url and parameters. It will select the default price shown on the screen which tends to be the lowest. Meta/Schema is all correct.

I have in excess of over 260 colours available in my swatch options. Not all used at the same time. They are a mixture of images and colours selected with the picker. A default exists.

**Comment by magento-engcom-team** [comment]
Hi @UKRogue. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@UKRogue do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
HI @UKRogue looks like this is duplicate to this one -> https://github.com/magento/magento2/issues/18017

**Comment by UKRogue** [comment]
https://github.com/magento/magento2/issues/18017 was concerning the changing of the image and the gallery loading at the wrong time to facilitate it. This issue, although similar is reference to the swatch parameter being actioned correctly, but then ignoring any additional parameters. The simple product image loads correctly, but seems to stop after that and not continue with anything else.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97126`, `MAGETWO-97127` were created

**Comment by vigna-sundar** [comment]
@engcom-backlog-andrii , I tried to reproduce the same issue in my community version 2.2.6 ..
It's working as expected I couldn't reproduce the issue mentioned here.

**Comment by UKRogue** [comment]
I still have issue in 2.2.6 (My Test site). I will turn off all extensions and try again just to be sure. Is it possible it's due to the amount of swatches available in product attributes? It's the only other thing that's hard to emulate. I have over 200 options. Is there anyway you can add 200 options (Mix of images and colors) for test without manually entering each one? Or perhaps I could do a bulk delete on my test site.

**Comment by UKRogue** [comment]
Additional. If I change the product attribute to dropdown instead of visual swatch, flush and clean cache, static and javascript files **it works properly**. As soon as I change back to swatches it stops working. Trouble is it deletes all uploaded swatches and selected colours . However, all swatches being blank and still not working does mean its not an uploaded image issue or mix of colour/images issue.

**Comment by UKRogue** [comment]
For now I have changed my live site to a drop down box to allow google merchant to pick up correct prices. However, initial problem still remains.

**Comment by lfritsche** [comment]
@UKRogue I'm facing the same issue, but were able to reduce the issue to the Google URL Parameter:
e.g.: #139=26?gclid=EAIaIQobChMIjfnZhMmC4AIVE6aaCh0iggWgEAkYAiABEgJcNfD_BwE does not work
#139=26 works -> so Magento has a problem with the gclid param.

Solved it by removing all behind the ? param:

```
        /**
         * Get default options values settings with either URL query parameters
         * @private
         */
        _getSelectedAttributes: function () {
            var hashIndex = window.location.href.indexOf('#'),
              selectedAttributes = {},
              params;

            if (hashIndex !== -1) {
                params = $.parseQuery(window.location.href.substr(hashIndex + 1));

                for (var attribute in params) {
                    params[attribute] = params[attribute].split('?')[0]; // remove all behind '?'
                }

                selectedAttributes = _.invert(_.mapObject(_.invert(params), function (attributeId) {
                    var attribute = this.options.jsonConfig.mappedAttributes[attributeId];
                    return attribute ? attribute.code : attributeId;
                }.bind(this)));
            }

            return selectedAttributes;
        },
```

**Comment by UKRogue** [comment]
Slightly different issue as my problem only occurred if a colour swatch was involved in the parameter and there was more than 1 parameter ie Height & Colour. Since I changed colour from swatch to drop down selection in the attributes I have no problem and parameters resolve correctly. However, it's still an issue that needs to be fixed

**Comment by lfritsche** [comment]
I'm using Magento 2.2.6 in this case and the example above also works for more than one swatch in my case. I've got two swatches: size & style. Removing an ending ?param solved the issue for me.
I can send you a PM with a live example and the swatch-renderer.js I'm using. `#139=28&140=36`

**Comment by engcom-Bravo** [comment]
Hello @UKRogue.

Thank you for contribution and collaboration!

We are not able to reproduce this issue on the lates `2.3-develop` branch by provided steps.
![swatches](https://user-images.githubusercontent.com/51680745/65580145-c88d7600-df81-11e9-92fc-dda0de2c367f.gif)


We are closing this issue due to _branch `2.2-develop` was closed and Magento no longer accepts pull requests and issues to the v2.2 release lines to focus all development efforts on v2.3._
 See [Accepted pull requests and ported code](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#rules) for more details

If you still faced this issue on `2.3` please create a new Issue with all required details according to [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)
